{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import pandas as pd\n",
    "import re\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.model_selection import train_test_split\n",
    "from scipy import io\n",
    "from scipy.sparse import csr_matrix\n",
    "from gears import PertData\n",
    "from tqdm import tqdm\n",
    "import requests\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Download and splitting norman data\n",
    "pert_data = PertData('norman') # specific saved folder\n",
    "pert_data.load(data_name = 'norman') # specific dataset name\n",
    "pert_data.prepare_split(split = 'simulation', seed = 1) # get data split with seed\n",
    "pert_data.get_dataloader(batch_size = 32, test_batch_size = 128) # prepare data loader"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_folder = \"..data/gene_perturb_data/norman\"\n",
    "data_path = os.path.join(data_folder, \"perturb_processed.h5ad\")\n",
    "adata_norman = sc.read(data_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "perturbed_genes = []\n",
    "\n",
    "for condition in adata_norman.obs[\"condition\"]:\n",
    "    if condition != \"ctrl\":\n",
    "        split_condition = condition.split(\"+\")\n",
    "        for gene in split_condition:\n",
    "            if gene not in perturbed_genes:\n",
    "                if gene != \"ctrl\":\n",
    "                    perturbed_genes.append(gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_protein_sequence(protein_id: str) -> str:\n",
    "    base_url = \"https://rest.ensembl.org\"\n",
    "    headers_json = {\"Content-Type\": \"application/json\"}\n",
    "\n",
    "    url = f\"{base_url}/sequence/id/{protein_id}?type=protein\"\n",
    "    r = requests.get(url, headers={\"Content-Type\": \"text/plain\"})\n",
    "    r.raise_for_status()\n",
    "    protein_seq = r.text.strip()\n",
    "    \n",
    "    return protein_seq "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_status(gene_id):\n",
    "    url = f\"https://rest.ensembl.org/sequence/id/{gene_id}\"\n",
    "\n",
    "    params = {\n",
    "        \"type\": \"protein\",\n",
    "        \"multiple_sequences\": 1\n",
    "    }\n",
    "    headers = {\n",
    "        \"Content-Type\": \"text/x-fasta\"\n",
    "    }\n",
    "\n",
    "    response = requests.get(url, params=params, headers=headers)\n",
    "    return response.ok\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def gene_to_ensembl(gene_name, species=\"human\"):\n",
    "\n",
    "    url = f\"https://mygene.info/v3/query\"\n",
    "    params = {\n",
    "        \"q\": gene_name,\n",
    "        \"species\": species,\n",
    "        \"fields\": \"ensembl.gene\",\n",
    "        \"size\": 1\n",
    "    }\n",
    "    \n",
    "    try:\n",
    "        response = requests.get(url, params=params)\n",
    "        response.raise_for_status()\n",
    "        results = response.json()\n",
    "        \n",
    "        if \"hits\" in results and results[\"hits\"]:\n",
    "            hit = results[\"hits\"][0]\n",
    "            ensembl = hit.get(\"ensembl\")\n",
    "            if isinstance(ensembl, list):\n",
    "                return ensembl[0].get(\"gene\")\n",
    "            elif isinstance(ensembl, dict):\n",
    "                return ensembl.get(\"gene\")\n",
    "        return None\n",
    "    except Exception as e:\n",
    "        print(f\"Error retrieving Ensembl ID for {gene_name}: {e}\")\n",
    "        return None\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_ensembl_dict = {}\n",
    "perturbed_ids = []\n",
    "for gene in perturbed_genes:\n",
    "    ens_id = gene_to_ensembl(gene)\n",
    "    perturbed_ids.append(ens_id)\n",
    "    gene_ensembl_dict[gene] = ens_id\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dict_path = os.path.join(data_folder, \"gene_ensembl_dict.pkl\")\n",
    "\n",
    "with open(dict_path, \"wb\") as f:\n",
    "    pickle.dump(gene_ensembl_dict, f)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "successes = []\n",
    "failures = []\n",
    "\n",
    "for gene_id in tqdm(perturbed_ids):\n",
    "    status = get_status(gene_id)\n",
    "    if status == True:\n",
    "        successes.append(gene_id)\n",
    "    else:\n",
    "        failures.append(gene_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_canonical_transcript_id(ensembl_gene_id: str) -> str:\n",
    "    base_url = \"https://rest.ensembl.org\"\n",
    "    headers_json = {\"Content-Type\": \"application/json\"}\n",
    "    \n",
    "    # Step 1: Get canonical transcript for the gene\n",
    "    url = f\"{base_url}/lookup/id/{ensembl_gene_id}?expand=1\"\n",
    "    r = requests.get(url, headers=headers_json)\n",
    "    r.raise_for_status()\n",
    "    gene_info = r.json()\n",
    "    \n",
    "    canonical_transcript_id = gene_info.get(\"canonical_transcript\")\n",
    "    #if canonical_transcript_id is None:\n",
    "        #raise ValueError(f\"No canonical transcript found for gene {ensembl_gene_id}\")\n",
    "\n",
    "    return canonical_transcript_id\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_protein_translation_id(canonical_transcript_id: str) -> str:\n",
    "    base_url = \"https://rest.ensembl.org\"\n",
    "    headers_json = {\"Content-Type\": \"application/json\"}\n",
    "\n",
    "    url = f\"{base_url}/lookup/id/{canonical_transcript_id}?expand=1\"\n",
    "    r = requests.get(url, headers=headers_json)\n",
    "    r.raise_for_status()\n",
    "    tx_info = r.json()\n",
    "    \n",
    "    protein_id = tx_info.get(\"Translation\", {}).get(\"id\")\n",
    "    #if protein_id is None:\n",
    "        #raise ValueError(f\"No protein translation found for transcript {canonical_transcript_id}\")\n",
    "\n",
    "    return protein_id\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_protein_sequence(protein_id: str) -> str:\n",
    "    base_url = \"https://rest.ensembl.org\"\n",
    "    headers_json = {\"Content-Type\": \"application/json\"}\n",
    "\n",
    "    url = f\"{base_url}/sequence/id/{protein_id}?type=protein\"\n",
    "    r = requests.get(url, headers={\"Content-Type\": \"text/plain\"})\n",
    "    r.raise_for_status()\n",
    "    protein_seq = r.text.strip()\n",
    "    \n",
    "    return protein_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_dict = {}\n",
    "\n",
    "successful_genes = []\n",
    "failure_genes = []\n",
    "\n",
    "#Maybe determine successes first\n",
    "for gene_id in tqdm(successes):\n",
    "    protein_seq = None\n",
    "    canon_transcript_id = get_canonical_transcript_id(gene_id)\n",
    "    ct_id_new = canon_transcript_id.split(\".\")[0]\n",
    "    \n",
    "    if ct_id_new is not None:\n",
    "        protein_id = get_protein_translation_id(ct_id_new)\n",
    "        \n",
    "    if protein_id is not None:\n",
    "        protein_seq = get_protein_sequence(protein_id)\n",
    "\n",
    "    if protein_seq is not None:\n",
    "        successful_genes.append(gene_id)\n",
    "        aa_dict[gene_id] = protein_seq\n",
    "    else:\n",
    "        failure_genes.append(gene_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "csv_path = os.path.join(data_folder, \"perturbed_genes.csv\")\n",
    "\n",
    "aa_df = pd.DataFrame.from_dict(aa_dict, orient=\"index\", columns=[\"aa_sequence\"])\n",
    "aa_df = aa_df.reset_index().rename(columns={\"index\": \"gene_id\"})\n",
    "aa_df.to_csv(csv_path, index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pinnacle3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
