{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from tqdm import tqdm\n",
    "import os\n",
    "\n",
    "import pickle\n",
    "import requests\n",
    "from sklearn.model_selection import train_test_split\n",
    "import torch\n",
    "import h5py\n",
    "import ast\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "from scipy.io import mmread\n",
    "\n",
    "from scipy.sparse import csr_matrix\n",
    "from gears import PertData\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_status(gene_id):\n",
    "    url = f\"https://rest.ensembl.org/sequence/id/{gene_id}\"\n",
    "\n",
    "    params = {\n",
    "        \"type\": \"protein\",\n",
    "        \"multiple_sequences\": 1\n",
    "    }\n",
    "    headers = {\n",
    "        \"Content-Type\": \"text/x-fasta\"\n",
    "    }\n",
    "\n",
    "    response = requests.get(url, params=params, headers=headers)\n",
    "    return response.ok\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_canonical_transcript_id(ensembl_gene_id: str) -> str:\n",
    "    base_url = \"https://rest.ensembl.org\"\n",
    "    headers_json = {\"Content-Type\": \"application/json\"}\n",
    "    \n",
    "    # Step 1: Get canonical transcript for the gene\n",
    "    url = f\"{base_url}/lookup/id/{ensembl_gene_id}?expand=1\"\n",
    "    r = requests.get(url, headers=headers_json)\n",
    "    r.raise_for_status()\n",
    "    gene_info = r.json()\n",
    "    \n",
    "    canonical_transcript_id = gene_info.get(\"canonical_transcript\")\n",
    "    #if canonical_transcript_id is None:\n",
    "        #raise ValueError(f\"No canonical transcript found for gene {ensembl_gene_id}\")\n",
    "\n",
    "    return canonical_transcript_id\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_protein_translation_id(canonical_transcript_id: str) -> str:\n",
    "    base_url = \"https://rest.ensembl.org\"\n",
    "    headers_json = {\"Content-Type\": \"application/json\"}\n",
    "\n",
    "    url = f\"{base_url}/lookup/id/{canonical_transcript_id}?expand=1\"\n",
    "    r = requests.get(url, headers=headers_json)\n",
    "    r.raise_for_status()\n",
    "    tx_info = r.json()\n",
    "    \n",
    "    protein_id = tx_info.get(\"Translation\", {}).get(\"id\")\n",
    "    #if protein_id is None:\n",
    "        #raise ValueError(f\"No protein translation found for transcript {canonical_transcript_id}\")\n",
    "\n",
    "    return protein_id\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_protein_sequence(protein_id: str) -> str:\n",
    "    base_url = \"https://rest.ensembl.org\"\n",
    "    headers_json = {\"Content-Type\": \"application/json\"}\n",
    "\n",
    "    url = f\"{base_url}/sequence/id/{protein_id}?type=protein\"\n",
    "    r = requests.get(url, headers={\"Content-Type\": \"text/plain\"})\n",
    "    r.raise_for_status()\n",
    "    protein_seq = r.text.strip()\n",
    "    \n",
    "    return protein_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Data processing for replogle RPE1 dataset\n",
    "\n",
    "data_folder_rpe1 = \"..data/gene_perturb_data/replogle_rpe1\"\n",
    "\n",
    "raw_data_path = os.path.join(data_folder_rpe1, \"rpe1_raw_singlecell_01.h5ad\")\n",
    "replogle_rpe1 = sc.read_h5ad(raw_data_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_ensembl_dict = {}\n",
    "\n",
    "used_genes = []\n",
    "for i, gene in enumerate(replogle_rpe1.obs[\"gene\"]):\n",
    "    if gene not in used_genes:\n",
    "        ensembl_id = replogle_rpe1.obs[\"gene_id\"][i]\n",
    "        gene_ensembl_dict[gene] = ensembl_id\n",
    "        used_genes.append(gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dict_path_rpe1 = os.path.join(data_folder_rpe1, \"gene_ensembl_dict.pkl\")\n",
    "\n",
    "with open(dict_path_rpe1, \"wb\") as f:\n",
    "    pickle.dump(gene_ensembl_dict, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.normalize_total(replogle_rpe1)\n",
    "sc.pp.log1p(replogle_rpe1)\n",
    "sc.pp.highly_variable_genes(replogle_rpe1,n_top_genes=5000, subset=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "perturbed_genes = list(replogle_rpe1.obs[\"gene_id\"].unique())\n",
    "perturbed_genes.remove('non-targeting')\n",
    "perturbed_genes.remove('nan')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#The successes are protein coding genes\n",
    "#The failures are either not found in the database or are non-coding genes\n",
    "\n",
    "successes = []\n",
    "failures = []\n",
    "\n",
    "for gene_id in tqdm(perturbed_genes):\n",
    "    status = get_status(gene_id)\n",
    "    if status == True:\n",
    "        successes.append(gene_id)\n",
    "    else:\n",
    "        failures.append(gene_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_dict = {}\n",
    "\n",
    "successful_genes = []\n",
    "failure_genes = []\n",
    "\n",
    "#Maybe determine successes first\n",
    "for gene_id in tqdm(successes):\n",
    "    protein_seq = None\n",
    "    canon_transcript_id = get_canonical_transcript_id(gene_id)\n",
    "    ct_id_new = canon_transcript_id.split(\".\")[0]\n",
    "    \n",
    "    if ct_id_new is not None:\n",
    "        protein_id = get_protein_translation_id(ct_id_new)\n",
    "        \n",
    "    if protein_id is not None:\n",
    "        protein_seq = get_protein_sequence(protein_id)\n",
    "\n",
    "    if protein_seq is not None:\n",
    "        successful_genes.append(gene_id)\n",
    "        aa_dict[gene_id] = protein_seq\n",
    "    else:\n",
    "        failure_genes.append(gene_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "csv_path_rpe1 = os.path.join(data_folder_rpe1, \"perturbed_genes.csv\")\n",
    "\n",
    "aa_df = pd.DataFrame.from_dict(aa_dict, orient=\"index\", columns=[\"aa_sequence\"])\n",
    "aa_df = aa_df.reset_index().rename(columns={\"index\": \"gene_id\"})\n",
    "aa_df.to_csv(csv_path_rpe1, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_rpe1 = replogle_rpe1.copy()\n",
    "filtered_rpe1.X = csr_matrix(filtered_rpe1.X)\n",
    "cell_type = 'rpe1'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_rpe1.obs = filtered_rpe1.obs.rename(columns={'gene':'condition'})\n",
    "filtered_rpe1.obs['condition'] = [c + '+ctrl' for c in filtered_rpe1.obs['condition']]\n",
    "filtered_rpe1.obs['cell_type'] = cell_type\n",
    "filtered_rpe1.obs = filtered_rpe1.obs.loc[:,['condition', 'cell_type']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set condition names\n",
    "mapper = {k:k for k in filtered_rpe1.obs['condition'].unique()}\n",
    "mapper['non-targeting+ctrl'] = 'ctrl'\n",
    "filtered_rpe1.obs['condition'] = filtered_rpe1.obs['condition'].map(mapper)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set cov_drug_dose_name names\n",
    "filtered_rpe1.obs['cov_drug_dose_name'] = [cell_type+'_'+x+'_1+1' for x in filtered_rpe1.obs['condition']]\n",
    "mapper = {k:k for k in filtered_rpe1.obs['cov_drug_dose_name'].unique()}\n",
    "mapper['rpe1_ctrl_1+1'] = 'rpe1_ctrl_1'\n",
    "filtered_rpe1.obs['cov_drug_dose_name'] = filtered_rpe1.obs['cov_drug_dose_name'].map(mapper)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from gears.data_utils import rank_genes_groups_by_cov\n",
    "\n",
    "rank_genes_groups_by_cov(filtered_rpe1, groupby='cov_drug_dose_name', \n",
    "                         covariate='cell_type', control_group='ctrl_1', n_genes=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_rpe1.X = csr_matrix(filtered_rpe1.X)\n",
    "adata_rpe1 = filtered_rpe1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#cwd = os.getcwd()\n",
    "pert_data = PertData(\"..data/gene_perturb_data\") # specific saved folder\n",
    "pert_data.new_data_process(dataset_name = data_folder_rpe1, adata = adata_rpe1) # specific dataset name and adata object\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pert_data.load(data_path = data_folder_rpe1) # load the processed data, the path is saved folder + dataset_name\n",
    "pert_data.prepare_split(split = 'simulation', seed = 1) # get data split with seed\n",
    "pert_data.get_dataloader(batch_size = 32, test_batch_size = 128) # prepare data loader"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Data processing for replogle K562 dataset\n",
    "\n",
    "data_folder_k562 = \"..data/gene_perturb_data/replogle_k562\"\n",
    "\n",
    "raw_data_path = os.path.join(data_folder_k562, \"K562_essential_raw_singlecell_01.h5ad\")\n",
    "replogle_k562 = sc.read_h5ad(raw_data_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_ensembl_dict = {}\n",
    "\n",
    "used_genes = []\n",
    "for i, gene in enumerate(replogle_k562.obs[\"gene\"]):\n",
    "    if gene not in used_genes:\n",
    "        ensembl_id = replogle_k562.obs[\"gene_id\"][i]\n",
    "        gene_ensembl_dict[gene] = ensembl_id\n",
    "        used_genes.append(gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dict_path_k562 = os.path.join(data_folder_k562, \"gene_ensembl_dict.pkl\")\n",
    "\n",
    "with open(dict_path_k562, \"wb\") as f:\n",
    "    pickle.dump(gene_ensembl_dict, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.normalize_total(replogle_k562)\n",
    "sc.pp.log1p(replogle_k562)\n",
    "\n",
    "sc.pp.highly_variable_genes(replogle_k562,n_top_genes=5000, subset=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "perturbed_genes = list(replogle_k562.obs[\"gene_id\"].unique())\n",
    "perturbed_genes.remove('non-targeting')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "successes = []\n",
    "failures = []\n",
    "\n",
    "for gene_id in tqdm(perturbed_genes):\n",
    "    status = get_status(gene_id)\n",
    "    if status == True:\n",
    "        successes.append(gene_id)\n",
    "    else:\n",
    "        failures.append(gene_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_dict = {}\n",
    "\n",
    "successful_genes = []\n",
    "failure_genes = []\n",
    "\n",
    "#Maybe determine successes first\n",
    "for gene_id in tqdm(successes):\n",
    "    protein_seq = None\n",
    "    canon_transcript_id = get_canonical_transcript_id(gene_id)\n",
    "    ct_id_new = canon_transcript_id.split(\".\")[0]\n",
    "    \n",
    "    if ct_id_new is not None:\n",
    "        protein_id = get_protein_translation_id(ct_id_new)\n",
    "        \n",
    "    if protein_id is not None:\n",
    "        protein_seq = get_protein_sequence(protein_id)\n",
    "\n",
    "    if protein_seq is not None:\n",
    "        successful_genes.append(gene_id)\n",
    "        aa_dict[gene_id] = protein_seq\n",
    "    else:\n",
    "        failure_genes.append(gene_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "csv_path_k562 = os.path.join(data_folder_k562, \"perturbed_genes.csv\")\n",
    "\n",
    "aa_df = pd.DataFrame.from_dict(aa_dict, orient=\"index\", columns=[\"aa_sequence\"])\n",
    "aa_df = aa_df.reset_index().rename(columns={\"index\": \"gene_id\"})\n",
    "aa_df.to_csv(csv_path_k562, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_k562 = replogle_k562.copy()\n",
    "filtered_k562.X = csr_matrix(filtered_k562.X)\n",
    "cell_type = 'K562'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_k562.obs = filtered_k562.obs.rename(columns={'gene':'condition'})\n",
    "filtered_k562.obs['condition'] = [c + '+ctrl' for c in filtered_k562.obs['condition']]\n",
    "filtered_k562.obs['cell_type'] = cell_type\n",
    "filtered_k562.obs = filtered_k562.obs.loc[:,['condition', 'cell_type']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set condition names\n",
    "mapper = {k:k for k in filtered_k562.obs['condition'].unique()}\n",
    "mapper['non-targeting+ctrl'] = 'ctrl'\n",
    "filtered_k562.obs['condition'] = filtered_k562.obs['condition'].map(mapper)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set cov_drug_dose_name names\n",
    "filtered_k562.obs['cov_drug_dose_name'] = [cell_type+'_'+x+'_1+1' for x in filtered_k562.obs['condition']]\n",
    "mapper = {k:k for k in filtered_k562.obs['cov_drug_dose_name'].unique()}\n",
    "mapper['K562_ctrl_1+1'] = 'K562_ctrl_1'\n",
    "filtered_k562.obs['cov_drug_dose_name'] = filtered_k562.obs['cov_drug_dose_name'].map(mapper)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from gears.data_utils import rank_genes_groups_by_cov\n",
    "\n",
    "rank_genes_groups_by_cov(filtered_k562, groupby='cov_drug_dose_name', \n",
    "                         covariate='cell_type', control_group='ctrl_1', n_genes=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_k562.X = csr_matrix(filtered_k562.X)\n",
    "adata_k562 = filtered_k562"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pert_data = PertData(\"..data/gene_perturb_data\") # specific saved folder\n",
    "pert_data.new_data_process(dataset_name = data_folder_k562, adata = adata_k562) # specific dataset name and adata object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pert_data.load(data_path = data_folder_k562) # load the processed data, the path is saved folder + dataset_name\n",
    "pert_data.prepare_split(split = 'simulation', seed = 1) # get data split with seed\n",
    "pert_data.get_dataloader(batch_size = 32, test_batch_size = 128) # prepare data loader"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pinnacle3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
