{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "import pickle\n",
    "import requests\n",
    "from sklearn.model_selection import train_test_split\n",
    "import torch\n",
    "import h5py\n",
    "import ast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Data processing for therapeutic target prediction task\n",
    "data_folder = \"..data/therapeutic_target_data\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_genes = []\n",
    "ppi_edgelist_path = os.path.join(data_folder, \"networks/global_ppi_edgelist.txt\")\n",
    "\n",
    "with open(ppi_edgelist_path, \"r\") as ppi_file:\n",
    "    for line in ppi_file:\n",
    "        genes = line.strip().split()\n",
    "        if genes[0] not in unique_genes:\n",
    "            unique_genes.append(genes[0])\n",
    "        if genes[1] not in unique_genes:\n",
    "            unique_genes.append(genes[1])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_uniprot_id(gene_symbol):\n",
    "    url = f\"https://rest.uniprot.org/uniprotkb/search?query=gene:{gene_symbol}&format=json\"\n",
    "    response = requests.get(url)\n",
    "    \n",
    "    if response.status_code == 200:\n",
    "        data = response.json()\n",
    "        if data['results']:\n",
    "            # Extract the first entry's UniProt ID\n",
    "            uniprot_id = data['results'][0]['primaryAccession']\n",
    "            return uniprot_id\n",
    "        else:\n",
    "            return \"No UniProt ID found for this gene symbol.\"\n",
    "    else:\n",
    "        return f\"Error: {response.status_code}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Mapping all of the genes to proteins\n",
    "protein_dict = {}\n",
    "\n",
    "for gene in tqdm(unique_genes):\n",
    "    uniprot_id = get_uniprot_id(gene)\n",
    "    protein_dict[gene] = uniprot_id\n",
    "    \n",
    "#Saving the dictionary that maps genes to proteins\n",
    "#with open(\"protein_dict.pkl\", \"wb\") as saved_file:\n",
    "    #pickle.dump(protein_dict, saved_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unmatched_genes = []\n",
    "\n",
    "for gene in list(protein_dict.keys()):\n",
    "    if protein_dict[gene] == \"No UniProt ID found for this gene symbol.\":\n",
    "        unmatched_genes.append(gene)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_amino_acid_sequence(uniprot_id):\n",
    "\n",
    "    url = f\"https://www.uniprot.org/uniprot/{uniprot_id}.fasta\"\n",
    "    response = requests.get(url)\n",
    "\n",
    "    if response.status_code == 200:\n",
    "        fasta_data = response.text\n",
    "        sequence = \"\".join(fasta_data.splitlines()[1:])  # Skip the first line (header)\n",
    "        return sequence\n",
    "    else:\n",
    "        return \"Error\"\n",
    "        #return f\"Error: Unable to fetch data for {uniprot_id}. Status code: {response.status_code}\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_dict = {}\n",
    "uniprot_ids = protein_dict.values()\n",
    "for uni_id in tqdm(uniprot_ids):\n",
    "    if uni_id != \"No UniProt ID found for this gene symbol.\":\n",
    "        aa_sequence = get_amino_acid_sequence(uni_id)\n",
    "        aa_dict[uni_id] = aa_sequence\n",
    "        \n",
    "#Saving the dictionary that maps proteins to amino acids\n",
    "#with open(\"aa_dict.pkl\", \"wb\") as saved_file:\n",
    "    #pickle.dump(aa_dict, saved_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Converting dictionary to dataframe:\n",
    "columns = [\"Gene\", \"AA Sequence\"]\n",
    "genes = [gene for gene in aa_dict.keys()]\n",
    "sequences = [sequence for sequence in aa_dict.values()]\n",
    "\n",
    "sequence_df = pd.DataFrame(list(zip(genes, sequences)), columns=columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "not_included = []\n",
    "for protein in all_proteins:\n",
    "    if protein not in aa_dict.keys():\n",
    "        not_included.append(protein)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Checking for duplicate proteins:\n",
    "unique_proteins = []\n",
    "duplicates_dict = {}\n",
    "\n",
    "for gene, protein in protein_dict.items():\n",
    "    if protein != \"No UniProt ID found for this gene symbol.\":\n",
    "        if protein in unique_proteins:\n",
    "            duplicates_dict[gene] = protein\n",
    "        else:\n",
    "            unique_proteins.append(protein)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Dictionary that maps genes to amino acids\n",
    "gene_aa_dict = {}\n",
    "for gene in unique_genes:\n",
    "    uniprot_id = protein_dict[gene]\n",
    "    if uniprot_id != \"No UniProt ID found for this gene symbol.\":\n",
    "        aa_sequence = aa_dict[uniprot_id]\n",
    "        gene_aa_dict[gene] = aa_sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Saving the dictionary that maps proteins to amino acids\n",
    "gene_aa_dict_path = os.path.join(data_folder, \"gene_aa_dict.pkl\")\n",
    "\n",
    "with open(gene_aa_dict_path, \"wb\") as saved_file:\n",
    "    pickle.dump(gene_aa_dict, saved_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "labels_dict_path = os.path.join(data_folder, \"pinnacle_embeds/pinnacle_labels_dict.txt\")\n",
    "\n",
    "with open(labels_dict_path, \"r\") as f:\n",
    "    labels_dict = f.read()\n",
    "labels_dict = labels_dict.replace(\"\\'\", \"\\\"\")\n",
    "labels_dict = json.loads(labels_dict)\n",
    "celltypes = [c for c in labels_dict[\"Cell Type\"] if c.startswith(\"CCI\")]\n",
    "celltype_dict = {ct.split(\"CCI_\")[1]: i for i, ct in enumerate(celltypes)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_names = []\n",
    "protein_celltypes = []\n",
    "for c, p in zip(labels_dict[\"Cell Type\"], labels_dict[\"Name\"]):\n",
    "    if c.startswith(\"BTO\") or c.startswith(\"CCI\") or c.startswith(\"Sanity\"): continue\n",
    "    protein_names.append(p)\n",
    "    protein_celltypes.append(c)\n",
    "\n",
    "proteins = pd.DataFrame.from_dict({\"target\": protein_names, \"cell type\": protein_celltypes})\n",
    "celltype_protein_dict = proteins.pivot_table(values=\"target\", index=\"cell type\", aggfunc={\"target\": list}).to_dict()[\"target\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "celltype_dict_path = os.path.join(data_folder, \"celltype_dict.pkl\")\n",
    "celltype_protein_dict_path = os.path.join(data_folder, \"celltype_protein_dict.pkl\")\n",
    "\n",
    "with open(celltype_dict_path, \"wb\") as saved_file:\n",
    "    pickle.dump(celltype_dict, saved_file)\n",
    "\n",
    "with open(celltype_protein_dict_path, \"wb\") as saved_file:\n",
    "    pickle.dump(celltype_protein_dict, saved_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mg_labels_dict_path = os.path.join(data_folder, \"pinnacle_embeds/pinnacle_mg_labels_dict.txt\")\n",
    "with open(mg_labels_dict_path, \"r\") as file:\n",
    "    mg_labels_dict = ast.literal_eval(file.read())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mg_labels = mg_labels_dict[\"Cell Type\"]\n",
    "cell_embeds = {}\n",
    "#156 is the number of different cell-contexts\n",
    "for i in range(156):\n",
    "    label = mg_labels[i].split(\"_\")[1]\n",
    "    embedding = mg_embeds[i]\n",
    "    cell_embeds[label] = embedding\n",
    "\n",
    "cell_embeds_path = os.path.join(data_folder, \"cell_embeds.pkl\")\n",
    "with open(cell_embeds_path, \"wb\") as saved_file:\n",
    "    pickle.dump(cell_embeds, saved_file)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pinnacle3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
